### PR TITLE
#164350412  Define Reaction Table

### PR DESCRIPTION
**Pull request by obasajujoshua31**
#### Description of Task to be completed?

- created like and dislike table for articles and comments
- created relationship between reactions table and users table
- created relationship between reactions table and article table
- created relationship between reactions table and comments table

 #### Any background context you want to provide?
Already existing user model, article model, comment model and category model

 #### What are the relevant pivotal tracker stories?

[#164350412](https://www.pivotaltracker.com/story/show/164350412)

### CODE REVIEW

File: server/models/comments.js [hunk 1]
```diff
@@ -32,7 +32,23 @@ const getCommentModel = (sequelize, DataTypes) => {
...
+      foreignKey: "articleId",
+      onDelete: "CASCADE"
+    });
+    Comment.hasMany(db.Reactions, {
+      scopes: {
+        commentLikes: {
+          include: [
+            {
+              model: db.Reactions,
+              where: { sourceType: 2 }
```
Comment (benfluleck) [inline server/models/comments.js]: Please assign this to a constant it is not good practice to hard code values that are non-descriptive. If someone new is looking at this code. He will be wondering what the number 2 is doing there

File: server/tests/controller/article/article.spec.js [hunk 2]
```diff
@@ -42,6 +42,32 @@ describe("CRUD Article Feature </api/v1/article>", done => {
     expect(res.body.success).to.be.true;
   });
 
+  it("should not create an article", async () => {
```
Comment (benfluleck) [inline server/tests/controller/article/article.spec.js]: This test is not descriptive enough. `should not create an article` does not tell me anything

File: server/tests/controller/article/article.spec.js [hunk 3]
```diff
@@ -42,6 +42,32 @@ describe("CRUD Article Feature </api/v1/article>", done => {
...
+      .send({ ...articleData, categoryId: "" })
+      .set({ Authorization: token });
+    expect(res.statusCode).to.equal(400);
+    expect(res.body.success).to.be.false;
+    expect(res.body.errorMessages.categoryId).to.equal(
+      "Please select a category"
+    );
+  });
+
+  it("should return 500 error for wrong token", async () => {
```
Comment (benfluleck) [inline server/tests/controller/article/article.spec.js]: This test description is not too dissimilar to the one after it, please redefine it

File: server/tests/controller/article/article.spec.js [hunk 4]
```diff
@@ -137,6 +163,19 @@ describe("CRUD Article Feature </api/v1/article>", done => {
     expect(res.body.success).to.be.false;
   });
 
+  it("should return 500 error, if token  is malformed", async () => {
```
Comment (benfluleck) [inline server/tests/controller/article/article.spec.js]: This test is not too dissimilar to the one before it but they are testing different routes 🤷‍♂️

File: server/tests/models/reactions.spec.js [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+      articleId,
+      userId,
+      sourceType: 1,
+      reactions: true
+    });
+    assert.instanceOf(reactionInstance, Reactions);
+    assert.lengthOf(Object.keys(reactionInstance.dataValues), "7");
+  });
+
+  it("should like an article  ", async () => {
```
Comment (benfluleck) [inline server/tests/models/reactions.spec.js]: Make this test a bit more descriptive. At the moment I can see; Reactions table model from the describe block in line 40 and should like an article which does not make sense.

File: server/models/articles.js [hunk 6]
```diff
@@ -83,6 +92,21 @@ const getArticleModel = (sequelize, DataTypes) => {
       target: "id",
       onDelete: "CASCADE"
     });
+    Article.hasMany(db.Reactions, {
+      scopes: {
+        articleReactions: {
+          include: [
+            {
+              model: db.Reactions,
+              where: { sourceType: ARTICLE }
```
Comment (benfluleck) [inline server/models/articles.js]: Is this ARTICLE ID, NUMBER OF ARTICLE or the whole ARTICLE OBJECT

File: server/models/comments.js [hunk 7]
```diff
@@ -32,7 +34,23 @@ const getCommentModel = (sequelize, DataTypes) => {
...
+      foreignKey: "articleId",
+      onDelete: "CASCADE"
+    });
+    Comment.hasMany(db.Reactions, {
+      scopes: {
+        commentLikes: {
+          include: [
+            {
+              model: db.Reactions,
+              where: { sourceType: COMMENT }
```
Comment (benfluleck) [inline server/models/comments.js]: same thing here please be a bit more descripitive

File: server/tests/models/reactions.spec.js [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+    const createdCommentLikes = await Reactions.create({
+      userId,
+      commentId,
+      reactions: true,
+      sourceType: 2
+    });
+    assert.equal(createdCommentLikes.dataValues.reactions, true);
+  });
+
+  it("should dislike a comment  ", async () => {
```
Comment (benfluleck) [inline server/tests/models/reactions.spec.js]: Please make this test more descriptive

File: server/tests/models/reactions.spec.js [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+    const createdArticleDislikes = await Reactions.create({
+      userId,
+      articleId,
+      reactions: false,
+      sourceType: 1
+    });
+    assert.equal(createdArticleDislikes.dataValues.reactions, false);
+  });
+
+  it("should like a comment  ", async () => {
```
Comment (benfluleck) [inline server/tests/models/reactions.spec.js]: Please make this test more descriptive

File: server/tests/models/reactions.spec.js [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+    const articleLikes = await Reactions.create({
+      userId,
+      articleId,
+      reactions: true,
+      sourceType: 1
+    });
+    assert.equal(articleLikes.dataValues.reactions, true);
+  });
+
+  it("should dislike an article  ", async () => {
```
Comment (benfluleck) [inline server/tests/models/reactions.spec.js]: Please make this test more descriptive

**Review by benfluleck — CHANGES_REQUESTED** [review]

File: server/tests/models/reactions.spec.js [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+    const articleLikes = await Reactions.create({
+      userId,
+      articleId,
+      reactions: true,
+      sourceType: ARTICLE_REACTION
+    });
+    assert.equal(articleLikes.dataValues.reactions, true);
+  });
+
+  it("should add a dislike reaction to an article  ", async () => {
```
Comment (benfluleck) [inline server/tests/models/reactions.spec.js]: get rid of the white spaces please

File: server/tests/models/reactions.spec.js [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+    const createdArticleDislikes = await Reactions.create({
+      userId,
+      articleId,
+      reactions: false,
+      sourceType: ARTICLE_REACTION
+    });
+    assert.equal(createdArticleDislikes.dataValues.reactions, false);
+  });
+
+  it("should add a like reaction to a comment  ", async () => {
```
Comment (benfluleck) [inline server/tests/models/reactions.spec.js]: get rid of the white spaces please

File: server/tests/models/reactions.spec.js [hunk 13]
```diff
@@ -0,0 +1,142 @@
...
+    const createdCommentLikes = await Reactions.create({
+      userId,
+      commentId,
+      reactions: true,
+      sourceType: COMMENT_REACTION
+    });
+    assert.equal(createdCommentLikes.dataValues.reactions, true);
+  });
+
+  it("should add a dislike reaction to a comment  ", async () => {
```
Comment (benfluleck) [inline server/tests/models/reactions.spec.js]: get rid of the white spaces please

**Review by benfluleck — CHANGES_REQUESTED** [review]

File: server/models/comments.js [hunk 14]
```diff
@@ -32,7 +34,23 @@ const getCommentModel = (sequelize, DataTypes) => {
     });
 
     Comment.belongsTo(db.Articles, {
-      foreignKey: "articleId"
+      foreignKey: "articleId",
+      onDelete: "CASCADE"
+    });
+    Comment.hasMany(db.Reactions, {
+      scopes: {
```
Comment (Ndifreke) [inline server/models/comments.js:41]: This looks like something out of NASA, I believe keeping simple things simple can make it much easy to follow. Is there any way we can make this more readable

File: server/tests/models/reactions.spec.js [hunk 15]
```diff
@@ -0,0 +1,142 @@
...
+
+  return Promise.resolve({
+    userId,
+    articleId,
+    commentId
+  });
+};
+
+describe("Reactions table model", () => {
+  it("should create a Reaction table with with a length of 7", async () => {
```
Comment (Ndifreke) [inline server/tests/models/reactions.spec.js:42]: How about 'should create Reactions with 7 properties/fields'?

**Review by Ndifreke — APPROVED** [review]

File: server/tests/controller/user/profile.spec.js [hunk 16]
```diff
@@ -92,7 +103,7 @@ describe("Test for User Profile", () => {
         .send(userData[2]);
       expect(res.statusCode).to.equal(200);
       expect(res.body.success).to.be.true;
-      expect(res.body.message).to.equal("User profile updated");
+      expect(res.body.message).to.eql("User profile updated");
```
Comment (ghost) [inline server/tests/controller/user/profile.spec.js:106]: why did you change this please

File: server/models/comments.js (hunk 14 above)
Comment (obasajujoshua31) [reply server/models/comments.js:41]: please check here for more information. https://sequelize.readthedocs.io/en/latest/docs/scopes/

**Review by ghost — APPROVED** [review]

File: server/tests/controller/user/profile.spec.js (hunk 16 above)
Comment (obasajujoshua31) [reply server/tests/controller/user/profile.spec.js:106]: I thought that eql is more appropriate for testing strings

**Review by benfluleck — APPROVED** [review]
